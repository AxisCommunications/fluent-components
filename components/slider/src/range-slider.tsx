import { ForwardRefComponent } from "@fluentui/react-utilities";
import React from "react";

import { renderSlider_unstable } from "./render-slider";
import { RangeSliderProps } from "./slider.types";
import { useSliderContextValues } from "./use-slider-context-values";
import { useSliderStyles_unstable } from "./use-slider-styles";
import { useRangeSlider_unstable } from "./use-range-slider";

export const RangeSlider: ForwardRefComponent<RangeSliderProps> = React
  .forwardRef((props, ref) => {
    const state = useRangeSlider_unstable(props, ref);
    const contextValues = useSliderContextValues(state);

    useSliderStyles_unstable(state);

    return renderSlider_unstable(state, contextValues);
  });

RangeSlider.displayName = "RangeSlider";
