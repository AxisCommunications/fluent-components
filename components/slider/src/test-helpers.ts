import { vi } from "vitest";
import { sliderClassNames } from "./use-slider-styles";

export const getControlRoot = (sliderRoot: HTMLElement): HTMLElement => {
  const element = sliderRoot.querySelector(`.${sliderClassNames.control}`);
  if (element === null) {
    throw new Error("Unable to find control element");
  }
  element.getBoundingClientRect = () => ({
    x: 0,
    y: 0,
    width: 100,
    height: 10,
    bottom: 10,
    left: 0,
    right: 0,
    top: 0,
    toJSON: vi.fn(),
  });
  return element as HTMLElement;
};
