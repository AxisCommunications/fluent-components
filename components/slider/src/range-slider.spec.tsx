import { fireEvent, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { RangeSlider } from "./range-slider";
import React from "react";
import { getControlRoot } from "./test-helpers";
import { describe, expect, it, vi } from "vitest";

const expectSliderValues = (elements: HTMLElement[], values: number[]) => {
  expect(elements).toHaveLength(values.length);
  elements.forEach((element, i) => {
    expect(element.getAttribute("value")).toEqual(values[i].toString());
  });
};

describe("range-slider", () => {
  it("should render", () => {
    const result = render(
      <RangeSlider value={[20, 40]} marks={[]} min={0} max={100} />
    );

    expect(result).toMatchSnapshot();
  });

  it("should apply value", () => {
    const { getAllByRole } = render(
      <RangeSlider min={0} max={100} value={[10, 20]} />
    );
    expectSliderValues(getAllByRole("slider"), [10, 20]);
  });

  it("should apply default value", () => {
    const { getAllByRole } = render(
      <RangeSlider min={0} max={100} defaultValue={[10, 20]} />
    );
    expectSliderValues(getAllByRole("slider"), [10, 20]);
  });

  describe("mouse interaction", () => {
    it("should set closest thumb to value", () => {
      const { getByTestId, getAllByRole } = render(
        <RangeSlider
          min={0}
          max={100}
          defaultValue={[10, 60]}
          data-testid="slider-root"
        />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.mouseDown(controlRoot, { button: 0, clientX: 50 });

      expectSliderValues(getAllByRole("slider"), [10, 50]);
    });

    it("should swap focused thumb", () => {
      const { getByTestId, getAllByRole } = render(
        <RangeSlider
          min={0}
          max={100}
          defaultValue={[40, 60]}
          data-testid="slider-root"
        />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.mouseDown(controlRoot, { button: 0, clientX: 40 });
      fireEvent.mouseMove(controlRoot, { buttons: 1, clientX: 70 });

      expectSliderValues(getAllByRole("slider"), [60, 70]);
    });

    it("should call onChange but not committed on down", () => {
      const onChange = vi.fn();
      const onChangeCommitted = vi.fn();
      const { getByTestId } = render(
        <RangeSlider
          min={0}
          max={100}
          defaultValue={[0, 100]}
          onChange={onChange}
          onChangeCommitted={onChangeCommitted}
          data-testid="slider-root"
        />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.mouseDown(controlRoot, { button: 0, clientX: 40 });

      expect(onChange).toHaveBeenCalledWith({ value: [40, 100] });
      expect(onChangeCommitted).not.toHaveBeenCalled();
    });

    it("should call onChange but not committed on move", () => {
      const onChange = vi.fn();
      const onChangeCommitted = vi.fn();
      const { getByTestId } = render(
        <RangeSlider
          min={0}
          max={100}
          defaultValue={[0, 100]}
          onChange={onChange}
          onChangeCommitted={onChangeCommitted}
          data-testid="slider-root"
        />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.mouseDown(controlRoot, { button: 0, clientX: 10 });
      fireEvent.mouseMove(controlRoot, { buttons: 1, clientX: 30 });

      expect(onChange).toHaveBeenCalledWith({ value: [30, 100] });
      expect(onChangeCommitted).not.toHaveBeenCalled();
    });

    it("should call onChangeCommitted", () => {
      const onChangeCommitted = vi.fn();
      const { getByTestId } = render(
        <RangeSlider
          min={0}
          max={100}
          defaultValue={[0, 100]}
          onChangeCommitted={onChangeCommitted}
          data-testid="slider-root"
        />
      );

      const controlRoot = getControlRoot(getByTestId("slider-root"));
      fireEvent.mouseDown(controlRoot, { button: 0, clientX: 40 });
      fireEvent.mouseUp(controlRoot, { button: 0, clientX: 40 });

      expect(onChangeCommitted).toHaveBeenCalledWith({ value: [40, 100] });
    });

    it("should snap to closest mark", () => {
      const onChange = vi.fn();
      const onChangeCommitted = vi.fn();
      const { getByTestId } = render(
        <RangeSlider
          title="Range Slider"
          min={0}
          max={100}
          onChange={onChange}
          onChangeCommitted={onChangeCommitted}
          marks={[
            { value: 20, label: <span data-testid="mark">20</span> },
            { value: 40, label: "40" },
            { value: 60, label: "60" },
          ]}
          data-testid="slider-root"
        />
      );

      const mark: HTMLElement = getByTestId("mark");
      getControlRoot(getByTestId("slider-root"));
      fireEvent.mouseDown(mark, { button: 0, clientX: 29 });
      fireEvent.mouseUp(mark, { button: 0, clientX: 29 });

      expect(onChange).toHaveBeenCalledWith({ value: [20] });
      expect(onChangeCommitted).toHaveBeenCalledWith({ value: [20] });
    });
  });

  describe("keyboard interaction", () => {
    it("should move focus between thumbs", async () => {
      const { getAllByRole } = render(
        <RangeSlider
          min={0}
          max={100}
          defaultValue={[25, 75]}
          data-testid="slider-root"
        />
      );

      const [left, right] = getAllByRole("slider");
      expect(document.activeElement).not.toEqual(left);
      expect(document.activeElement).not.toEqual(right);

      await userEvent.tab();

      expect(document.activeElement).toEqual(left);

      await userEvent.tab();

      expect(document.activeElement).toEqual(right);
    });

    it("should swap focused thumb", () => {
      const { getAllByRole } = render(
        <RangeSlider min={0} max={100} defaultValue={[49, 50]} />
      );

      const sliders = getAllByRole("slider");
      const [left, right] = sliders;
      fireEvent.focus(left);
      fireEvent.input(left, { target: { valueAsNumber: 50 } });
      fireEvent.input(left, { target: { valueAsNumber: 51 } });

      expect(document.activeElement).toEqual(right);

      expectSliderValues(sliders, [50, 51]);
    });
  });
});
