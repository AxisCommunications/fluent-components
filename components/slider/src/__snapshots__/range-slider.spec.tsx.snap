// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`range-slider > should render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="axis-Slider ___16lm6va_1th6fmc fp2oml8 fifp7yv fwrwuds f1asdtw4 f4aqd28 f22iagw f10pi13n fly5x3f f3er7iy f1lx44mx fke05om ftqa4ok f2hkw1w f1b1k54r f4ne723 fh7aioi fqqcjud f1ufm4qn f1qnwcb4 fi52z01 fgrk5zm ffht0p2 f1p0ul1q f1c901ms f1alokd7 fmj8fco f1iwowo3 f1pffoy2 f1bmyog6 fs6b7xr f15fv2gd"
      >
        <span
          class="axis-Slider__control ___144j0yg_julj5g0 f10pi13n f14t3ns0 f1l02sjl fcgxt0o f1ujusj6 fz0d3q9 f1k6fduh fdwl3z2"
        >
          <span
            class="axis-Slider__rail ___kotjr40_1djcq1g fnivh3a fc7yr5o f1el4m67 f8yange f1p9o1ba f1sil6mw ftgm304 f1euv43f fly5x3f f1erzq44 f1i1t8d1 f188r07x f1d3ncnn"
          >
            <span
              class="axis-Slider__track ___pq61zr0_1xft6bt fnivh3a fc7yr5o f1el4m67 f8yange ftgm304 f1euv43f f1erzq44 f1i1t8d1 f188r07x f1yrba31 f174n9yu"
              style="width: 20%; left: 20%;"
            />
          </span>
          <div
            class="axis-Slider__thumb ___14hb9vm_jmfl5w0 f8fbkgy f1nfllo7 f1djnp8u f1s8kh49 f22iagw f4d9j23 f122n59 f1euv43f f1t2h5zl f4aqd28 f1i1t8d1 f1vcxb5z fk1lali f1ewtqcl frcgkvu feh19e8 f3j3guj fo75joz f1yag6sl fsc6avr f1w15cjc fkaq55s f1fp5o0y f163fonl f1yq6w5o f11yjt3y f1jpmc5p f10tv6oz f16xp3sf fwrmqbx f1seuxxq f1j7ml58 f14u7mkt f5zrw40 fto0uou f1ks5ppg fyl8oag f1wl9k8s f1mdlcz9 fmmikit f10awi5f f1mfomsq fftmtmu f13zj6fq fgougqc fprarqb f14vs0nd f1gtfqs9 f18zvfd9 f8sy9bk fufffdp f1sclg23 f132a78j fcetbih f1kt6s4o"
            style="left: 20%;"
          >
            <input
              data-index="0"
              id="slider-thumb-r0"
              max="100"
              min="0"
              type="range"
              value="20"
            />
            <span
              class="axis-Slider__thumb-label ___1b1bz8r_1eqf7oo fk116j8 fg8tfde fpvge99 fv71qf3 f1ywm7hm fw5db7e f14wxoun f1uw59to f1aa9q02 f16jpd5f f1jar5jt fyu767a f1euv43f feml9b3 f22iagw f122n59 f4d9j23 fxugw4r f18ygdci fxeb0a7 f1j7ml58 f1wl9k8s f15w9qf9 f1js41m7 fguqkdo f1ljr5q2 fto0uou f1hr5hl2 f1brpl1h"
            >
              20
            </span>
          </div>
          <div
            class="axis-Slider__thumb ___14hb9vm_jmfl5w0 f8fbkgy f1nfllo7 f1djnp8u f1s8kh49 f22iagw f4d9j23 f122n59 f1euv43f f1t2h5zl f4aqd28 f1i1t8d1 f1vcxb5z fk1lali f1ewtqcl frcgkvu feh19e8 f3j3guj fo75joz f1yag6sl fsc6avr f1w15cjc fkaq55s f1fp5o0y f163fonl f1yq6w5o f11yjt3y f1jpmc5p f10tv6oz f16xp3sf fwrmqbx f1seuxxq f1j7ml58 f14u7mkt f5zrw40 fto0uou f1ks5ppg fyl8oag f1wl9k8s f1mdlcz9 fmmikit f10awi5f f1mfomsq fftmtmu f13zj6fq fgougqc fprarqb f14vs0nd f1gtfqs9 f18zvfd9 f8sy9bk fufffdp f1sclg23 f132a78j fcetbih f1kt6s4o"
            style="left: 40%;"
          >
            <input
              data-index="1"
              id="slider-thumb-r1"
              max="100"
              min="0"
              type="range"
              value="40"
            />
            <span
              class="axis-Slider__thumb-label ___1b1bz8r_1eqf7oo fk116j8 fg8tfde fpvge99 fv71qf3 f1ywm7hm fw5db7e f14wxoun f1uw59to f1aa9q02 f16jpd5f f1jar5jt fyu767a f1euv43f feml9b3 f22iagw f122n59 f4d9j23 fxugw4r f18ygdci fxeb0a7 f1j7ml58 f1wl9k8s f15w9qf9 f1js41m7 fguqkdo f1ljr5q2 fto0uou f1hr5hl2 f1brpl1h"
            >
              40
            </span>
          </div>
        </span>
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="axis-Slider ___16lm6va_1th6fmc fp2oml8 fifp7yv fwrwuds f1asdtw4 f4aqd28 f22iagw f10pi13n fly5x3f f3er7iy f1lx44mx fke05om ftqa4ok f2hkw1w f1b1k54r f4ne723 fh7aioi fqqcjud f1ufm4qn f1qnwcb4 fi52z01 fgrk5zm ffht0p2 f1p0ul1q f1c901ms f1alokd7 fmj8fco f1iwowo3 f1pffoy2 f1bmyog6 fs6b7xr f15fv2gd"
    >
      <span
        class="axis-Slider__control ___144j0yg_julj5g0 f10pi13n f14t3ns0 f1l02sjl fcgxt0o f1ujusj6 fz0d3q9 f1k6fduh fdwl3z2"
      >
        <span
          class="axis-Slider__rail ___kotjr40_1djcq1g fnivh3a fc7yr5o f1el4m67 f8yange f1p9o1ba f1sil6mw ftgm304 f1euv43f fly5x3f f1erzq44 f1i1t8d1 f188r07x f1d3ncnn"
        >
          <span
            class="axis-Slider__track ___pq61zr0_1xft6bt fnivh3a fc7yr5o f1el4m67 f8yange ftgm304 f1euv43f f1erzq44 f1i1t8d1 f188r07x f1yrba31 f174n9yu"
            style="width: 20%; left: 20%;"
          />
        </span>
        <div
          class="axis-Slider__thumb ___14hb9vm_jmfl5w0 f8fbkgy f1nfllo7 f1djnp8u f1s8kh49 f22iagw f4d9j23 f122n59 f1euv43f f1t2h5zl f4aqd28 f1i1t8d1 f1vcxb5z fk1lali f1ewtqcl frcgkvu feh19e8 f3j3guj fo75joz f1yag6sl fsc6avr f1w15cjc fkaq55s f1fp5o0y f163fonl f1yq6w5o f11yjt3y f1jpmc5p f10tv6oz f16xp3sf fwrmqbx f1seuxxq f1j7ml58 f14u7mkt f5zrw40 fto0uou f1ks5ppg fyl8oag f1wl9k8s f1mdlcz9 fmmikit f10awi5f f1mfomsq fftmtmu f13zj6fq fgougqc fprarqb f14vs0nd f1gtfqs9 f18zvfd9 f8sy9bk fufffdp f1sclg23 f132a78j fcetbih f1kt6s4o"
          style="left: 20%;"
        >
          <input
            data-index="0"
            id="slider-thumb-r0"
            max="100"
            min="0"
            type="range"
            value="20"
          />
          <span
            class="axis-Slider__thumb-label ___1b1bz8r_1eqf7oo fk116j8 fg8tfde fpvge99 fv71qf3 f1ywm7hm fw5db7e f14wxoun f1uw59to f1aa9q02 f16jpd5f f1jar5jt fyu767a f1euv43f feml9b3 f22iagw f122n59 f4d9j23 fxugw4r f18ygdci fxeb0a7 f1j7ml58 f1wl9k8s f15w9qf9 f1js41m7 fguqkdo f1ljr5q2 fto0uou f1hr5hl2 f1brpl1h"
          >
            20
          </span>
        </div>
        <div
          class="axis-Slider__thumb ___14hb9vm_jmfl5w0 f8fbkgy f1nfllo7 f1djnp8u f1s8kh49 f22iagw f4d9j23 f122n59 f1euv43f f1t2h5zl f4aqd28 f1i1t8d1 f1vcxb5z fk1lali f1ewtqcl frcgkvu feh19e8 f3j3guj fo75joz f1yag6sl fsc6avr f1w15cjc fkaq55s f1fp5o0y f163fonl f1yq6w5o f11yjt3y f1jpmc5p f10tv6oz f16xp3sf fwrmqbx f1seuxxq f1j7ml58 f14u7mkt f5zrw40 fto0uou f1ks5ppg fyl8oag f1wl9k8s f1mdlcz9 fmmikit f10awi5f f1mfomsq fftmtmu f13zj6fq fgougqc fprarqb f14vs0nd f1gtfqs9 f18zvfd9 f8sy9bk fufffdp f1sclg23 f132a78j fcetbih f1kt6s4o"
          style="left: 40%;"
        >
          <input
            data-index="1"
            id="slider-thumb-r1"
            max="100"
            min="0"
            type="range"
            value="40"
          />
          <span
            class="axis-Slider__thumb-label ___1b1bz8r_1eqf7oo fk116j8 fg8tfde fpvge99 fv71qf3 f1ywm7hm fw5db7e f14wxoun f1uw59to f1aa9q02 f16jpd5f f1jar5jt fyu767a f1euv43f feml9b3 f22iagw f122n59 f4d9j23 fxugw4r f18ygdci fxeb0a7 f1j7ml58 f1wl9k8s f15w9qf9 f1js41m7 fguqkdo f1ljr5q2 fto0uou f1hr5hl2 f1brpl1h"
          >
            40
          </span>
        </div>
      </span>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
